ibis/bigquery/udf/udf.py

```diff
@@ -0,0 +1,820 @@
+                this.width = width;
+                this.height = height;
+            }
+            get area() {
+                return (this.width * this.height);
+            }
+            perimeter() {
+                return (2 * (this.width + this.height));
+            }
+        }
+        return new Rectangle(width, height);
+    }
+    return my_rectangle(width, height);
+    """;
+    '''
+    def wrapper(f):
+        if not callable(f):
+            raise TypeError('f must be callable, got {}'.format(f))
+
+        udf_node = type(
+            f.__name__,
+            (BigQueryUDFNode,),
+            dict(input_type=input_type, output_type=output_type.array_type),
+        )
+
+        @compiles(udf_node)
+        def compiles_udf_node(t, expr):
+            return '{}({})'.format(
+                udf_node.__name__,
+                ', '.join(map(t.translate, expr.op().args))
+            )
+
+        source = PythonToJavaScriptTranslator(f).compile()
+        js = '''\
+CREATE TEMPORARY FUNCTION {name}({signature})
+RETURNS {return_type}
+LANGUAGE js AS """
+{strict}{source}
+return {name}({args});
+""";'''.format(
+            name=f.__name__,
+            return_type=ibis_type_to_bigquery_type(output_type),
+            source=source,
+            signature=', '.join(
+                '{name} {type}'.format(
+                    name=name,
+                    type=ibis_type_to_bigquery_type(type)
+                ) for name, type in zip(
+                   inspect.signature(f).parameters.keys(), input_type
+                )
+            ),
+            strict=repr('use strict') + ';\n' if strict else '',
+            args=', '.join(inspect.signature(f).parameters.keys())
+        )
+
+        @functools.wraps(f)
+        def wrapped(*args, **kwargs):
+            node = udf_node(*args, **kwargs)
+            node.js = js
+            return node.to_expr()
+
+        wrapped.__signature__ = inspect.signature(f)
+        wrapped.js = js
+
+        return wrapped
+    return wrapper
+
+
+def semicolon(f):
+    """Add a semicolon to the result of a visit_* call.
+
+    Parameters
+    ----------
+    f : callable
+    """
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
+        return f(*args, **kwargs) + ';'
+    return wrapper
+
+
+@rewrite.register(ast.Call(func=ast.Name(id='print')))
+def rewrite_print(node):
+    return ast.Call(
+        func=ast.Attribute(
+            value=ast.Name(id='console', ctx=ast.Load()),
+            attr='log',
+            ctx=ast.Load()
+        ),
+        args=node.args,
+        keywords=node.keywords,
+    )
+
+
+@rewrite.register(ast.Call(func=ast.Attribute(attr='append')))
+def rewrite_append(node):
+    return ast.Call(
+        func=ast.Attribute(
+            value=node.func.value,
+            attr='push',
+            ctx=ast.Load(),
+        ),
+        args=node.args,
+        keywords=node.keywords,
+    )
+
+
+class PythonToJavaScriptTranslator:
```

Comment:
How about moving this to `transpiler.py` along with the `find` utility in `find.py`? Find is only used during translation (maybe `Rewriter` too).